### PR TITLE
[build-script] Add swift-disable-dead-stripping option for disabling dead stripping

### PR DESCRIPTION
**Pull request by mininny**
With the new build script flag `--swift-disable-dead-stripping`, swift frontend will be built without dead stripping which can improve debugging. 

Resolves [SR-13631](https://bugs.swift.org/browse/SR-13631).

### CODE REVIEW

File: cmake/modules/AddSwift.cmake [hunk 1]
```diff
@@ -384,8 +384,10 @@ function(_add_host_variant_link_flags target)
     if(CMAKE_SYSTEM_NAME MATCHES Darwin)
       # See rdar://48283130: This gives 6MB+ size reductions for swift and
       # SourceKitService, and much larger size reductions for sil-opt etc.
-      target_link_options(${target} PRIVATE
-        "SHELL:-Xlinker -dead_strip")
+      if (NOT SWIFT_DISABLE_DEAD_STRIPPING)
+        target_link_options(${target} PRIVATE
+          "SHELL:-Xlinker -dead_strip")
+      endif()
```
Comment (varungandhi-apple) [inline cmake/modules/AddSwift.cmake:390]: You could simplify the surrounding code here using the [`if(<cond1> AND <cond2>)` form](https://cmake.org/cmake/help/latest/command/if.html#condition-syntax).

File: utils/build_swift/build_swift/driver_arguments.py [hunk 2]
```diff
@@ -334,6 +334,9 @@ def create_argument_parser():
            default=defaults.SWIFT_ANALYZE_CODE_COVERAGE,
            help='enable code coverage analysis in Swift (false, not-merged, '
                 'merged).')
+                
+    option('--swift-disable-dead-stripping', toggle_true, 
+           help="Disable dead stripping in Swift build")
```
Comment (varungandhi-apple) [inline utils/build_swift/build_swift/driver_arguments.py:339]: For the help text, I think we should be more specific, say something like "Turn off Darwin-specific dead stripping for swiftc and other tools."

Comment (mininny) [reply utils/build_swift/build_swift/driver_arguments.py]: Thanks! I changed it to "Turn off Darwin-specific dead stripping for Swift host tools." as you suggested :)

**Review by varungandhi-apple — APPROVED** [review]

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test and merge

**Comment by varungandhi-apple** [comment]
Couple of Python test failures:

```
******************** TEST 'Swift(macosx-x86_64) :: Python/build_swift.swift' FAILED ********************
Script:
--
: 'RUN: at line 10';   /Applications/Xcode.app/Contents/Developer/usr/bin/python3 /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/build_swift/run_tests.py
--
Exit Code: 1

Command Output (stderr):
--
.....................................................................................................................................................................................................................................................................................................................F.........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
======================================================================
FAIL: test_expected_options_exhaustive (tests.build_swift.test_driver_arguments.TestDriverArgumentParser)
Test that we are exhaustively testing all options accepted by the
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/build_swift/tests/build_swift/test_driver_arguments.py", line 418, in test_expected_options_exhaustive
    self.fail('non-exhaustive expected options, missing: {}'
AssertionError: non-exhaustive expected options, missing: {'--swift-disable-dead-stripping'}

----------------------------------------------------------------------
Ran 1039 tests in 33.278s

FAILED (failures=1)

--

********************
PASS: Swift(macosx-x86_64) :: Evolution/test_enum_add_cases_trap.swift (7486 of 14077)
PASS: Swift(macosx-x86_64) :: Evolution/test_conformance_reference.swift (7487 of 14077)
PASS: Swift(macosx-x86_64) :: Evolution/test_enum_add_cases.swift (7488 of 14077)
FAIL: Swift(macosx-x86_64) :: Python/python_lint.swift (7489 of 14077)
******************** TEST 'Swift(macosx-x86_64) :: Python/python_lint.swift' FAILED ********************
Script:
--
: 'RUN: at line 8';   /Applications/Xcode.app/Contents/Developer/usr/bin/python3 /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/python_lint.py
--
Exit Code: 1

Command Output (stdout):
--
./utils/build-script:603:1: W293 blank line contains whitespace
./utils/build_swift/build_swift/driver_arguments.py:337:1: W293 blank line contains whitespace

--

********************
```

**Comment by mininny** [comment]
> Couple of Python test failures:
> 
> ```
> ******************** TEST 'Swift(macosx-x86_64) :: Python/build_swift.swift' FAILED ********************
> Script:
> --
> : 'RUN: at line 10';   /Applications/Xcode.app/Contents/Developer/usr/bin/python3 /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/build_swift/run_tests.py
> --
> Exit Code: 1
> 
> Command Output (stderr):
> --
> .....................................................................................................................................................................................................................................................................................................................F.........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
> ======================================================================
> FAIL: test_expected_options_exhaustive (tests.build_swift.test_driver_arguments.TestDriverArgumentParser)
> Test that we are exhaustively testing all options accepted by the
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/build_swift/tests/build_swift/test_driver_arguments.py", line 418, in test_expected_options_exhaustive
>     self.fail('non-exhaustive expected options, missing: {}'
> AssertionError: non-exhaustive expected options, missing: {'--swift-disable-dead-stripping'}
> 
> ----------------------------------------------------------------------
> Ran 1039 tests in 33.278s
> 
> FAILED (failures=1)
> 
> --
> 
> ********************
> PASS: Swift(macosx-x86_64) :: Evolution/test_enum_add_cases_trap.swift (7486 of 14077)
> PASS: Swift(macosx-x86_64) :: Evolution/test_conformance_reference.swift (7487 of 14077)
> PASS: Swift(macosx-x86_64) :: Evolution/test_enum_add_cases.swift (7488 of 14077)
> FAIL: Swift(macosx-x86_64) :: Python/python_lint.swift (7489 of 14077)
> ******************** TEST 'Swift(macosx-x86_64) :: Python/python_lint.swift' FAILED ********************
> Script:
> --
> : 'RUN: at line 8';   /Applications/Xcode.app/Contents/Developer/usr/bin/python3 /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/utils/python_lint.py
> --
> Exit Code: 1
> 
> Command Output (stdout):
> --
> ./utils/build-script:603:1: W293 blank line contains whitespace
> ./utils/build_swift/build_swift/driver_arguments.py:337:1: W293 blank line contains whitespace
> 
> --
> 
> ********************
> ```

Thanks for letting me know! I tried to fix it by looking around the code, but I couldn't test it with validation test on my MacBook(it keeps getting stuck). Can you please try run the test again? :)

**Comment by varungandhi-apple** [comment]
Sure. Btw, one way your should be able to run them is by something like:

```
$ utils/build-script [... other flags ...] --validation-test
# Hit Ctrl-C once it starts executing the validation tests
# This should create a validation-test-macosx-x86_64 directory
/path/to/lit.py -s -vv /path/to/swift-macosx-x86_64/validation-test-macosx-x86_64 --filter "Python"
```

@swift-ci smoke test

**Comment by mininny** [comment]
> Sure. Btw, one way your should be able to run them is by something like:
> 
> ```
> $ utils/build-script [... other flags ...] --validation-test
> # Hit Ctrl-C once it starts executing the validation tests
> # This should create a validation-test-macosx-x86_64 directory
> /path/to/lit.py -s -vv /path/to/swift-macosx-x86_64/validation-test-macosx-x86_64 --filter "Python"
> ```
> 
> @swift-ci smoke test

Thanks! :) Can you please trigger the test again?

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test

**Pull request merged**